File: tests/unit/TaskTests.cpp
#include <QtTest>
#include <QSignalSpy>
#include "domain/Task.h"
#include "domain/Correction.h"
#include "domain/ChecklistItem.h"

class TaskTests : public QObject
{
    Q_OBJECT

private slots:
    void testDefaults();
    void testNameProperty();
    void testDateProperty();
    void testAddItem();
    void testRemoveItem();
    void testAddCorrection();
    void testRemoveCorrection();
    void testSignals();
};

void TaskTests::testDefaults()
{
    Task task;
    QVERIFY(!task.id().isEmpty());
    QCOMPARE(task.name(), QString("Sin nombre"));
    QVERIFY(!task.date().isEmpty());
    QCOMPARE(task.itemCount(), 0);
    QCOMPARE(task.correctionCount(), 0);
}

void TaskTests::testNameProperty()
{
    Task task("Mi Tarea");
    QCOMPARE(task.name(), QString("Mi Tarea"));

    task.setName("Actualizada");
    QCOMPARE(task.name(), QString("Actualizada"));
}

void TaskTests::testDateProperty()
{
    QDate today(2024, 6, 20);
    Task task("Test", today);
    QCOMPARE(task.date(), QString("2024-06-20"));
}

void TaskTests::testAddItem()
{
    Task task;
    task.addNewItem("Item 1");
    task.addNewItem("Item 2");

    QCOMPARE(task.itemCount(), 2);
    QCOMPARE(task.getItem(0)->text(), QString("Item 1"));
    QCOMPARE(task.getItem(1)->text(), QString("Item 2"));
}

void TaskTests::testRemoveItem()
{
    Task task;
    task.addNewItem("Item 1");
    task.addNewItem("Item 2");

    task.removeItem(0);
    QCOMPARE(task.itemCount(), 1);
    QCOMPARE(task.getItem(0)->text(), QString("Item 2"));
}

void TaskTests::testAddCorrection()
{
    Task task;
    Correction *corr = task.addNewCorrection("Corrección 1");

    QCOMPARE(task.correctionCount(), 1);
    QVERIFY(corr != nullptr);
    QCOMPARE(corr->name(), QString("Corrección 1"));
}

void TaskTests::testRemoveCorrection()
{
    Task task;
    task.addNewCorrection("Corr 1");
    task.addNewCorrection("Corr 2");

    task.removeCorrection(0);
    QCOMPARE(task.correctionCount(), 1);
    QCOMPARE(task.getCorrection(0)->name(), QString("Corr 2"));
}

void TaskTests::testSignals()
{
    Task task;
    QSignalSpy nameSpy(&task, &Task::nameChanged);
    QSignalSpy itemCountSpy(&task, &Task::itemCountChanged);
    QSignalSpy corrCountSpy(&task, &Task::correctionCountChanged);

    task.setName("New name");
    QCOMPARE(nameSpy.count(), 1);

    task.addNewItem("Item");
    QCOMPARE(itemCountSpy.count(), 1);

    task.addNewCorrection("Corr");
    QCOMPARE(corrCountSpy.count(), 1);
}

QTEST_MAIN(TaskTests)
#include "TaskTests.moc"

File: csharp/tests/Checklist.Core.Tests/TaskTests.cs
using Checklist.Domain;
using Xunit;

namespace Checklist.Core.Tests;

public class TaskTests
{
    [Fact]
    public void Defaults()
    {
        var task = new Task();
        Assert.False(string.IsNullOrEmpty(task.Id));
        Assert.Equal("Unnamed", task.Name);
        Assert.False(string.IsNullOrEmpty(task.Date.ToString("yyyy-MM-dd")));
        Assert.Equal(0, task.ItemCount);
        Assert.Equal(0, task.CorrectionCount);
    }

    [Fact]
    public void NameProperty()
    {
        var task = new Task("Mi Tarea");
        Assert.Equal("Mi Tarea", task.Name);

        task.Name = "Actualizada";
        Assert.Equal("Actualizada", task.Name);
    }

    [Fact]
    public void DateProperty()
    {
        var today = new DateOnly(2024, 6, 20);
        var task = new Task("Test", today);
        Assert.Equal("2024-06-20", task.Date.ToString("yyyy-MM-dd"));
    }

    [Fact]
    public void AddItem()
    {
        var task = new Task();
        task.AddNewItem("Item 1");
        task.AddNewItem("Item 2");

        Assert.Equal(2, task.ItemCount);
        Assert.Equal("Item 1", task.GetItem(0)?.Text);
        Assert.Equal("Item 2", task.GetItem(1)?.Text);
    }

    [Fact]
    public void RemoveItem()
    {
        var task = new Task();
        task.AddNewItem("Item 1");
        task.AddNewItem("Item 2");

        task.RemoveItem(0);
        Assert.Equal(1, task.ItemCount);
        Assert.Equal("Item 2", task.GetItem(0)?.Text);
    }

    [Fact]
    public void AddCorrection()
    {
        var task = new Task();
        var corr = task.AddNewCorrection("Corrección 1");

        Assert.Equal(1, task.CorrectionCount);
        Assert.NotNull(corr);
        Assert.Equal("Corrección 1", corr.Name);
    }

    [Fact]
    public void RemoveCorrection()
    {
        var task = new Task();
        task.AddNewCorrection("Corr 1");
        task.AddNewCorrection("Corr 2");

        task.RemoveCorrection(0);
        Assert.Equal(1, task.CorrectionCount);
        Assert.Equal("Corr 2", task.GetCorrection(0)?.Name);
    }

    [Fact]
    public void Signals()
    {
        var task = new Task();
        var nameCount = 0;
        var itemCountCount = 0;
        var corrCountCount = 0;
        task.NameChanged += () => nameCount++;
        task.ItemCountChanged += () => itemCountCount++;
        task.CorrectionCountChanged += () => corrCountCount++;

        task.Name = "New name";
        Assert.Equal(1, nameCount);

        task.AddNewItem("Item");
        Assert.Equal(1, itemCountCount);

        task.AddNewCorrection("Corr");
        Assert.Equal(1, corrCountCount);
    }

    [Fact]
    public void CategoryProperty()
    {
        var task = new Task("Tarea");
        Assert.Equal("", task.Category);

        var changedCount = 0;
        task.CategoryChanged += () => changedCount++;

        task.Category = "Proyecto A";
        Assert.Equal("Proyecto A", task.Category);
        Assert.Equal(1, changedCount);

        task.Category = null!;
        Assert.Equal("", task.Category);
        Assert.Equal(2, changedCount);
    }

    [Fact]
    public void IsFinished()
    {
        var task = new Task("Tarea");
        Assert.False(task.IsFinished);
        Assert.False(task.IsFinished);

        task.AddNewItem("Item 1");
        task.AddNewItem("Item 2");
        Assert.False(task.IsFinished);

        task.GetItem(0)!.Done = true;
        Assert.False(task.IsFinished);

        task.GetItem(1)!.Done = true;
        Assert.True(task.IsFinished);
    }
}